Lib/dataclasses.py

```diff
@@ -1202,7 +1202,14 @@ def _get_slots(cls):
         # A class which does not define __slots__ at all is equivalent
         # to a class defining __slots__ = ('__dict__', '__weakref__')
         case None:
-            yield from ('__dict__', '__weakref__')
+            # Except for special cases, inheriting from them do not set
+            # any slots at all:
+            slots = ['__dict__', '__weakref__']
+            if getattr(cls, '__weakrefoffset__', -1) == 0:
+                slots.remove('__weakref__')
+            if getattr(cls, '__dictrefoffset__', -1) == 0:
+                slots.remove('__dict__')
```

Comment:
It seems a bit clearer to me to build this up rather than removing items from it:
```suggestion
            slots = []
            if getattr(cls, '__weakrefoffset__', -1) != 0:
                slots.append('__weakref__')
            if getattr(cls, '__dictrefoffset__', -1) != 0:
                slots.append('__dict__')
```